soccer/gameplay/tactics/defense2.py

```diff
@@ -0,0 +1,352 @@
+
+        self.assign_handlers_to_threats(
+            assigned_handlers, unused_threat_handlers, threats_to_block)
+
+        self.set_defender_block_lines(threats_to_block, assigned_handlers)
+
+    ## Gets list of threats
+    #  @return tuple of threat positions and score (unordered)
+    def get_threat_list(self, unused_threat_handlers):
+        # List of (position, score)
+        threats = []
+        potential_threats = main.their_robots()
+
+        # find the primary threat
+        # if the ball is not moving OR it's moving towards our goal, it's the primary threat
+        # if it's moving, but not towards our goal, the primary threat is the robot on their team most likely to catch it
+        if (main.ball().vel.mag() > 0.4):
+            if evaluation.ball.is_moving_towards_our_goal():
+                # Add tuple of pos and score
+                threats.append((main.ball().pos, 1))
+            else:
+                # Get all potential receivers
+                potential_receivers = []
+                for opp in potential_threats:
+                    if self.estimate_potential_recievers_score(opp):
+                        potential_receivers.append((opp.pos, 1))
+
+                if len(potential_receivers) > 0:
+                    # Add best receiver to threats
+                    # TODO Calc shot chance
+                    best_tuple = min(potential_receivers,
+                                     key=lambda rcrv_tuple: rcrv_tuple[1])
+                    threats.append((best_tuple[0], .81))
+                else:
+                    # Just deal with ball if no recievers
+                    threats.append((main.ball().pos, .9))
+        else:
+            # Assume opp is dribbling ball
+            if not constants.Field.OurGoalZoneShape.contains_point(
+                    main.ball().pos):
+                # TODO: Calc shot chance
+                threats.append((main.ball().pos, 1))
+
+        # if there are threats, check pass and shot chances
+        # If the first item is not a ball, it is most likely a pass
+        if len(threats) > 0 and threats[0][0] != main.ball().pos:
+            for opp in potential_threats:
+
+                # Exclude robots that have been assigned already
+                self.kick_eval.excluded_robots.clear()
+                for r in map(lambda bhvr: bhvr.robot, unused_threat_handlers):
+                    self.kick_eval.add_excluded_robot(r)
+
+                threats.append((opp.pos, self.estimate_risk_score(opp)))
+        else:
+            for opp in potential_threats:
+
+                # Exclude all robots
+                self.kick_eval.excluded_robots.clear()
+                for r in main.our_robots():
+                    self.kick_eval.add_excluded_robot(r)
+
+                shotChance = self.kick_eval.eval_pt_to_our_goal(opp.pos)
+
+                # Note: 0.5 is a bullshit value
+                threats.append((opp.pos, 0.5 * shotChance))
```

Comment:
This is left over from the other file. What I think is happening is that when a pass has already been kicked, just assume that it will succeed with 50% chance and just defend based on the shot chance of each robot. This allows us to block the higher shot chance opponent robots ignoring all other factors. If we base it off of the pass chance from the moving ball to other robots, the defenders will almost always track the ball as it moves across the field. This is due to the pass chance being heavily dependent on the distance from the ball to the receiving robot.

It's more of a stylistic decision on which direction we go.